### PR TITLE
Allow selection of shapes in extract_shape

### PR DESCRIPTION
**Pull request by jvegasbsc**
I was using the extract_shape preprocessor with the IHO Sea Areas shapefile from https://marineregions.org/downloads.php and realized that we have no way to specify  a subset of shapes in the file to use. I added it, along with a couple of tweaks to the code to make it work:

- This file has alphanumeric IDs, so I remove the conversion to int
- Anyway, I changed the way we assign the ID from the attributes to prioritize the attribute `name` if present.
- The `flat` call was failing for me, so I replaced it with `np.ravel` to create a  new view of the array

A sample of the new call:
```yaml
    extract_shape:
      shapefile: /home/Earth/jvegas/World_Seas_IHO_v3/World_Seas_IHO_v3.shp
      decomposed: True
      method: contains
      crop: False
      ids:
        - Arctic Ocean
        - Southern Ocean
        - Mediterranean Sea - Eastern Basin
        - Mediterranean Sea - Western Basin
        - North Atlantic Ocean
        - South Atlantic Ocean
        - North Pacific Ocean
        - South Pacific Ocean
        - Indian Ocean
```

I also tested the changes with the Natural Earth file for country boundaries and worked nice:

```yaml
    extract_shape:
      shapefile: /home/Earth/jvegas/NaturalEarth/Countries/ne_110m_admin_0_countries.shp
      decomposed: True
      method: contains
      crop: False
      ids:
        - Spain
        - France
        - Italy
        - Taiwan
```

**Tasks**

-   [ ] [Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do, if you haven't done so already (and add the link at the bottom)
-   [x] This pull request has a descriptive title that can be used in a changelog
-   [ ] Add unit tests
-   [x] Public functions should have a numpy-style docstring so they appear properly in the [API documentation](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/api/esmvalcore.html). For all other functions a one line docstring is sufficient.
-   [x] If writing a new/modified preprocessor function, please update the [documentation](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/recipe/preprocessor.html)
-   [x] Circle/CI tests pass. Status can be seen below your pull request. If the tests are failing, click the link to find out why.
-   [x] Codacy code quality checks pass. Status can be seen below your pull request. If there is an error, click the link to find out why. If you suspect Codacy may be wrong, please ask by commenting.
-   [x] Please use `yamllint` to check that your YAML files do not contain mistakes
-   [x] If you make backward incompatible changes to the recipe format, make a new pull request in the [ESMValTool repository](https://github.com/ESMValGroup/ESMValTool) and add the link below

If you need help with any of the tasks above, please do not hesitate to ask by commenting in the issue or pull request.

* * *

Closes #issue_number

Related to #758

### CODE REVIEW

**Comment by valeriupredoi** [comment]
something that @ledm would be happy to see, cheers @jvegasbsc :beer:

**Review by valeriupredoi — CHANGES_REQUESTED** [review]
would you be able to write a test for the new id members @jvegasbsc - it'd be nice to have this in the release :+1:

**Comment by jvegasbsc** [comment]
Working on the tests...

File: doc/recipe/preprocessor.rst [hunk 1]
```diff
@@ -1084,6 +1084,11 @@ Parameters:
     regions in the shape file is masked out. If ``true``, the regions in the
     shapefiles are masked out separately, generating an auxiliary dimension
     for the cube for this.
+  * ``ids``: by default, ``[]``, in this case all the shapes in the file will
+    be used. If a list of IDs is provided, only the shapes matching them will
+    be used. The IDs are assigned from the ``name`` or ``id`` attributes (in
+    that order of priority) ifpresent in the file or from the reading order
```
Comment (valeriupredoi) [inline doc/recipe/preprocessor.rst:1090]:
```suggestion
    that order of priority) if present in the file or from the reading order
```

File: doc/recipe/preprocessor.rst [hunk 2]
```diff
@@ -1084,6 +1084,11 @@ Parameters:
...
+    if not.
```
Comment (valeriupredoi) [inline doc/recipe/preprocessor.rst:1091]:
```suggestion
    if otherwise not present.
```

**Review by valeriupredoi — APPROVED** [review]
good stuffs @jvegasbsc :beer:

**Comment by valeriupredoi** [comment]
actually, looking at the documentation a bit closer maybe it'd be nice to have the examples from your PR description in as well - gives the user a clear picture what they can extract:
```yaml
    extract_shape:
      shapefile: /home/Earth/jvegas/World_Seas_IHO_v3/World_Seas_IHO_v3.shp
      decomposed: True
      method: contains
      crop: False
      ids:
        - Arctic Ocean
        - Southern Ocean
        - Mediterranean Sea - Eastern Basin
        - Mediterranean Sea - Western Basin
        - North Atlantic Ocean
        - South Atlantic Ocean
        - North Pacific Ocean
        - South Pacific Ocean
        - Indian Ocean

I also tested the changes with the Natural Earth file for country boundaries and worked nice:

    extract_shape:
      shapefile: /home/Earth/jvegas/NaturalEarth/Countries/ne_110m_admin_0_countries.shp
      decomposed: True
      method: contains
      crop: False
      ids:
        - Spain
        - France
        - Italy
       -  United Kingdom (does it work with spaces?)
        - Taiwan
```

File: doc/recipe/preprocessor.rst [hunk 3]
```diff
@@ -1084,6 +1084,11 @@ Parameters:
     regions in the shape file is masked out. If ``true``, the regions in the
     shapefiles are masked out separately, generating an auxiliary dimension
     for the cube for this.
+  * ``ids``: by default, ``[]``, in this case all the shapes in the file will
+    be used. If a list of IDs is provided, only the shapes matching them will
+    be used. The IDs are assigned from the ``name`` or ``id`` attributes (in
+    that order of priority) if present in the file or from the reading order
```
Comment (JaroCamphuijsen) [inline doc/recipe/preprocessor.rst:1090]: I understand from the code that it assigns numerical id's in reading order if the `name` and `id` attributes are not present. This could potentially result in unwanted and unpredictable behavior so it should be very clear in the documentation. Maybe an example that shows how this works might help. Currently it is only covered in a couple of words which are not completely clear.

**Review by JaroCamphuijsen — CHANGES_REQUESTED** [review]
@jvegasbsc great that you want to add this functionality. I left a couple of comments for improvement. We should also test whether the recipes that are currently using the preprocessor still work.

**Comment by JaroCamphuijsen** [comment]
@jvegasbsc Any plans to continue on this? I think it would be great to merge this!

**Comment by jvegasbsc** [comment]
Sorry, yes! I will update it now

File: doc/recipe/preprocessor.rst (hunk 3 above)
Comment (jvegasbsc) [reply doc/recipe/preprocessor.rst:1203]: Updated

**Comment by JaroCamphuijsen** [comment]
@jvegasbsc Great, I think your last update improved on the points I mentioned, however you also introduced some Codacy issues. Can you solve these? Then I would be ok with merging.

**Review by JaroCamphuijsen — APPROVED** [review]

**Comment by JaroCamphuijsen** [comment]
Actually many of the codacy issues are indentation issues. Also if we decide that the many variables are not a problem in ESMValCore, we should changes this in the Codacy settings.

**Comment by jvegasbsc** [comment]
> if we decide that the many variables are not a problem in ESMValCore, we should changes this in the Codacy settings.

As I see it, it is usually a problem, but in some specific cases we may want to ignore it: if there is no clear way to split the method or to extract parts of it to a new function, if we create many variables to make a later statement easier to understand.

By the way, this is not a case in which we should let it pass

**Comment by jvegasbsc** [comment]
The last codacy issues I think we can not solve.

**Comment by valeriupredoi** [comment]
cheers @jvegasbsc and @JaroCamphuijsen :+1: good work all round!
